src/lib.rs

```diff
@@ -1,19 +1,521 @@
+    {
+        self.statement = Some(statement.into());
+        self
+    }
+
+    fn pick_domain(&mut self) {
+        self.domain = match (self.domain.take(), &self.uri) {
+            (Some(d), _) => Some(d),
+            (None, Some(u)) => u
+                .as_str()
+                .parse::<Uri>()
+                .ok()
+                .as_ref()
+                .and_then(Uri::authority)
+                .cloned(),
+            (None, None) => None,
+        };
+    }
+
+    pub fn builder(mut self) -> Result<Builder, BuildError> {
+        self.pick_domain();
+
+        let uri = BuildError::require("uri", self.uri)?.to_string();
+
+        Ok(Builder {
+            domain: BuildError::require("domain", self.domain)?,
+            statement: self.statement,
+            uri: UriString::try_from(uri).expect("`Uri` must be a valid `UriString`"),
+            chain_id: BuildError::require("chain_id", self.chain_id)?,
+
+            nonce: Default::default(),
+            address: Default::default(),
+            issued_at: Default::default(),
+            expiration_time: Default::default(),
+            not_before: Default::default(),
+            request_id: Default::default(),
+            resources: Default::default(),
+        })
+    }
+
+    pub fn verifier<S>(mut self, message: S) -> Result<Verifier, VerifyError>
+    where
+        S: AsRef<str>,
+    {
+        self.pick_domain();
+
+        let message: Message = message.as_ref().parse()?;
+
+        let uri = VerifyError::require("uri", self.uri)?.to_string();
+
+        Ok(Verifier {
+            message,
+            valid_at: Default::default(),
+
+            domain: VerifyError::require("domain", self.domain)?,
+
+            statement: self.statement,
+            statement_mode: Verify::Always,
+
+            expiration_time_mode: Verify::WhenPresent,
+
+            not_before_mode: Verify::WhenPresent,
+
+            uri: UriString::try_from(uri).expect("`Uri` must be a valid `UriString`"),
+            chain_id: VerifyError::require("chain_id", self.chain_id)?,
+        })
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum Verify {
+    Always,
+    WhenPresent,
+    Never,
+}
+
+impl Verify {
+    fn verify<Expect, Actual, Cmp>(
+        self,
+        expect: Option<Expect>,
+        actual: Option<Actual>,
+        cmp: Cmp,
+    ) -> bool
+    where
+        Cmp: FnOnce(Expect, Actual) -> bool,
+    {
+        match (self, expect, actual) {
+            (Self::Never, _, _) => true,
+
+            (Self::Always, None, None) => true,
+            (Self::Always, Some(expect), Some(actual)) => cmp(expect, actual),
+            (Self::Always, _, _) => false,
+
+            (Self::WhenPresent, _, None) => true,
+            (Self::WhenPresent, None, Some(_)) => false,
+            (Self::WhenPresent, Some(expect), Some(actual)) => cmp(expect, actual),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+#[must_use]
+pub struct Verifier {
```

Comment:
This should simply be options passed to the existing verify method

Reply:
Using a method chain like this allows you to add new checks without having to make a semver incompatible change.

Reply:
I think @sbihel was saying that you can do the same by passing a `VerifyOpts` struct to the existing verify method, which also would allow you to add new checks without making a semver incompatible change.

Reply:
So perhaps something in the [`std::fs::OpenOptions`](https://doc.rust-lang.org/std/fs/struct.OpenOptions.html) style then?

```rust
let msg: Message = todo!();

// Using the defaults:
msg.verify()?;

// Customizing the options:
VerifyOptions::new()
    .not_before(Verify::Always)
    .expiration_time(Verify::Never)
    .statement(Verify::WhenPresent)
    .verify(&msg)?;
```